calc_from_args_utils/calculation_methods.py

```diff
@@ -0,0 +1,531 @@
+            input_vol = fetch_from_arg(art_tuple, args.vol_in, process)
+            if args.conc_units_in:
+                input_conc_units = str(
+                    fetch_from_arg(art_tuple, args.conc_units_in, process)
+                )
+                assert input_conc_units in [
+                    "ng/ul",
+                    "nM",
+                ], f'Unsupported conc. units "{input_conc_units}" for art {art_in.name}'
+            else:
+                # Infer concentration unit
+                if "ng/ul" in args.conc_in["udf"]:
+                    input_conc_units = "ng/ul"
+                elif "nM" in args.conc_in["udf"]:
+                    input_conc_units = "nM"
+                else:
+                    raise AssertionError(
+                        f"Can't infer units from '{args.conc_in['udf']}' for {art_out.name}."
+                    )
+                logging.info(
+                    f"Inferred unit of UDF '{args.conc_in['udf']}': {input_conc_units}."
+                )
+
+            # Infer amount unit
+            if "fmol" in args.amt_out["udf"]:
+                output_amt_unit = "fmol"
+            elif "ng" in args.amt_out["udf"]:
+                output_amt_unit = "ng"
+            else:
+                raise AssertionError(
+                    f"Can't infer units from '{args.amt_out['udf']}' for art {art_out.name}"
+                )
+            logging.info(
+                f"Inferred unit of UDF '{args.amt_out['udf']}': {output_amt_unit}."
+            )
+
+            # Calculate amount
+            if input_conc_units == "nM":
+                if output_amt_unit == "fmol":
+                    output_amt = input_vol * input_conc
+                elif output_amt_unit == "ng":
+                    output_amt = formula.fmol_to_ng(input_vol * input_conc, size_bp)
+            elif input_conc_units == "ng/ul":
+                if output_amt_unit == "fmol":
+                    output_amt = formula.ng_to_fmol(input_vol * input_conc, size_bp)
+                elif output_amt_unit == "ng":
+                    output_amt = input_vol * input_conc
+            logging.info(
+                f"Calculating amount: {input_vol} ul of {input_conc} {input_conc_units} at {size_bp} bp -> {output_amt:.2f} {output_amt_unit}"
+            )
+
+            # Update amount UDF of output artifact
+            udf_tools.put(
+                target=get_UDF_source(art_tuple, args.amt_out, process),
+                target_udf=args.amt_out["udf"],
+                val=round(output_amt, 2),
+            )
+            logging.info(
+                f"Assigned UDF '{args.amt_out['udf']}': {output_amt:.2f} for {args.amt_out['source']} '{get_UDF_source_name(art_tuple, args.amt_out, process)}'."
+            )
+
+        except AssertionError as e:
+            logging.error(str(e), exc_info=True)
+            logging.warning("Skipping.")
+            continue
```

Comment:
Won't this lead to a "Script finished successfully" in the log in calc_from_args.py?

Reply:
It should be caught [here](https://github.com/SciLifeLab/scilifelab_epps/pull/321/files/8eabf2372d2a6d4a9a2f4a8d25e3adbb8aaa852a#diff-9599aa3feb4778835f434a02ab919e1da58b069c9e5f123159f0720a7563e9acR213)